Docker/Dockerfile

```diff
@@ -1,102 +1,201 @@
+FROM build_base AS build_common
 
-# Use conda-pack to create a standalone environment in /venv:
-RUN conda-pack -n fastsurfer_gpu -o /tmp/env.tar && \
-  mkdir /venv && cd /venv && tar xf /tmp/env.tar && \
-  rm /tmp/env.tar
-
-# Now that venv in a new location, fix up paths:
-RUN /venv/bin/conda-unpack
-ENV PATH /venv/bin:$PATH
+# get install scripts into docker
+COPY ./env/fastsurfer.yml ./Docker/conda_pack.sh ./Docker/install_env.py /install/
+
+# SHELL ["/bin/bash", "--login", "-c"]
+# Install conda for gpu
+
+ARG DEBUG=false
+RUN python /install/install_env.py -m base -i /install/fastsurfer.yml -o /install/base-env.yml && \
+    conda env create -f "/install/base-env.yml" | tee /install/env-create.log ; \
+    if [ "${DEBUG}" != "true" ]; then \
+      rm /install/base-env.yml ; \
+    fi
+
+FROM build_common AS build_conda
+
+ARG DEBUG=false
+ARG DEVICE=cu118
+# install additional packages for cuda/rocm/cpu
+RUN python /install/install_env.py -m ${DEVICE} -i /install/fastsurfer.yml -o /install/${DEVICE}-env.yml && \
+    conda env update -n "fastsurfer" -f "/install/${DEVICE}-env.yml" | tee /install/env-update.log && \
+    /install/conda_pack.sh "fastsurfer" && \
+    echo "DEBUG=$DEBUG\nDEVICE=$DEVICE\n" > /install/build_conda.args ;  \
+    if [ "${DEBUG}" != "true" ]; then \
+      conda env remove -n "fastsurfer" && \
+      rm -R /install ; \
+    fi
+
+# create a stage for pruned Freesurfer
+FROM build_base AS build_freesurfer
 
-# setup shell for install command below
-RUN echo "source /venv/bin/activate" >> ~/.bashrc
+# get install scripts into docker
+COPY ./Docker/install_fs_pruned.sh /install/
 SHELL ["/bin/bash", "--login", "-c"]
 
 # install freesurfer and point to new python location
-RUN /fastsurfer/install_fs_pruned.sh /opt --upx && \
-    rm /opt/freesurfer/bin/fspython && \
+RUN /install/install_fs_pruned.sh /opt --upx && \
+    rm /opt/freesurfer/bin/fspython &&  \
+    rm -R /install && \
     ln -s /venv/bin/python3 /opt/freesurfer/bin/fspython
 
-# ========================================
-# Here we create the smaller runtime image
-# ========================================
 
-FROM ubuntu:22.04 AS runtime
+# =======================================================
+# Here, we create references to the requested build image
+# =======================================================
+# This is needed because COPY --from=<image/stage> does not accept variables as part of the image/stage name
+# selected_freesurfer_build_image -> $FREESURFER_BUILD_IMAGE
+FROM $FREESURFER_BUILD_IMAGE AS selected_freesurfer_build_image
+# selected_conda_build_image -> $CONDA_BUILD_IMAGE
+FROM $CONDA_BUILD_IMAGE AS selected_conda_build_image
+
+
+# =========================================
+# Here, we create the smaller runtime image
+# =========================================
+FROM $RUNTIME_BASE_IMAGE AS runtime
 
-# Install required packages for freesurfer to run
+ENV LANG=C.UTF-8 \
+    HSA_OVERRIDE_GFX_VERSION=10.3.0
```

Comment:
maybe we need a comment here, why this is needed? And when it would need to be changed to a new version?

Reply:
This is a flag for AMD ROCm. It should probably not be part of the "standard" docker.
In fact, since this is a flag that enables ROCm support on hardware that AMD does not officially support ROCm for, it should probably not be part of the docker container at all.
I suggest this flag **has to be passed** by the user `docker run ... -e HSA_OVERRIDE_GFX_VERSION=10.3.0 fastsurfer:rocm ...` instead of blanket-setting it here. Because other hardware might need this variable to be "unset".